zebra-consensus/src/checkpoint.rs

```diff
@@ -0,0 +1,158 @@
+    ///
+    /// Checkpoints should be chosen to avoid forks or chain reorganizations,
+    /// which only happen in the last few hundred blocks in the chain.
+    /// (zcashd allows chain reorganizations up to 99 blocks, and prunes
+    /// orphaned side-chains after 288 blocks.)
+    checkpoint_list: Arc<HashMap<BlockHeight, BlockHeaderHash>>,
+}
+
+/// The error type for the CheckpointVerifier Service.
+// TODO(jlusby): Error = Report ?
+type Error = Box<dyn error::Error + Send + Sync + 'static>;
+
+/// The CheckpointVerifier service implementation.
+///
+/// After verification, blocks are added to the underlying state service.
+impl<S> Service<Arc<Block>> for CheckpointVerifier<S>
+where
+    S: Service<zebra_state::Request, Response = zebra_state::Response, Error = Error>
+        + Send
+        + Clone
+        + 'static,
+    S::Future: Send + 'static,
+{
+    type Response = BlockHeaderHash;
+    type Error = Error;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // We don't expect the state to exert backpressure on verifier users,
+        // so we don't need to call `state_service.poll_ready()` here.
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, block: Arc<Block>) -> Self::Future {
+        // TODO(jlusby): Error = Report, handle errors from state_service.
+        let mut state_service = self.state_service.clone();
+        let checkpoint_list = self.checkpoint_list.clone();
+
+        async move {
+            if checkpoint_list.is_empty() {
+                return Err("the checkpoint list is empty".into());
+            };
+
+            let block_height = match block.coinbase_height() {
+                Some(height) => height,
+                None => return Err("the block does not have a coinbase height".into()),
+            };
+
+            // TODO(teor):
+            //   - implement chaining from checkpoints to their ancestors
+            //   - if chaining is expensive, move this check to the Future
+            //   - should the state contain a mapping from previous_block_hash to block?
+            let checkpoint_hash = match checkpoint_list.get(&block_height) {
+                Some(&hash) => hash,
+                None => return Err("the block's height is not a checkpoint height".into()),
+            };
+
+            // Hashing is expensive, so we do it as late as possible
+            if BlockHeaderHash::from(block.as_ref()) != checkpoint_hash {
+                // The block is on a side-chain
+                return Err("the block hash does not match the checkpoint hash".into());
+            }
+
+            // `Tower::Buffer` requires a 1:1 relationship between `poll()`s
+            // and `call()`s, because it reserves a buffer slot in each
+            // `call()`.
+            // TODO(teor): what happens if the await fails?
```

Comment:
If `ready_and().await` resolves to an `Err` (signaling failure of the state service), the error will be bubbled up by `?` and verification of the block will fail with the error from the state service, i.e., block verification won't succeed unless the block can be written back to the state, which I think is what we want?

Reply:
(That said, I think that this service shouldn't be committing anything to the state, just returning a verification result, so that we can have separation of concerns and factor out the state-commitment behavior from this checkpoint verifier on the one hand and from the full verifier on the other hand).

Reply:
So the buffer slot will be used up, regardless of whether the error is in the buffer or state service?

(I'll factor out the state commitment once we have a HybridVerifier service.)

Reply:
I guess if the error is in the buffer, it won't reserve a slot. And if the error is in the state service, then it will use up the buffer slot.